### ISSUE TITLE
Portable Version [Windows]

### ISSUE THREAD

**Issue by smaragdus** [crow-translate/crow-translate#74]
### Preamble
After using [QTranslate](https://quest-app.appspot.com/) (desktop program) and **Translate Man** (browser extension- [Chromium](https://chrome.google.com/webstore/detail/translate-man/fapgabkkfcaejckbfmfcdgnfefbmlion), [Firefox](https://addons.mozilla.org/en-US/firefox/addon/%E7%BF%BB%E8%AF%91%E4%BE%A0-translate-man/)) for quite some time, several days ago I came across [Crow Translate](https://crow-translate.github.io/) (on [AlternativeTo](https://alternativeto.net/software/crow-translate/)). **QTranslate** is a well-designed and mature program but it is neither cross platform nor open source. As sooner or later I will have to switch from Windows to Linux (Windows 8 will be my last Windows version as I have not even considered updating to spyware Windows 10) I decided to have a look at **Crow Translate**. Screenshots and programming language usually can tell a lot so even before testing **Crow Translate** I supposed I would like the program- native (written in C++, no .NET, no Java, no abominations like JavaScript/Electron), cross platform and open source, clean interface, decent customization options. Testing **Crow Translate** just confirmed my expectations- the program is a keeper. Yet it does not (yet) have one of the virtues of **QTranslate**- portable version.

### Request
As I mentioned above **QTranslate** has a portable version- one needs just to download the archive, extract it, run the executable and all settings will be saved in a configuration file (_Options.json_) in a sub-folder (_Data_) inside program folder. As the portable version of **QTranslate** does not write neither to Windows Registry nor to AppData the user can easily move the program from machine to machine with all its settings.

Currently **Crow Translate** saves settings to Windows Registry:

`HKEY_CURRENT_USER\Software\Crow\Crow Translate`

The most important about portable version of a program (or a program running in portable mode) is that all settings are saved in a configuration file (_Config_, _INI_, _JSON_, _XML_, etc) inside program folder and nothing is written neither in AppData nor in Windows Registry. Portability can be implemented in different ways, I will provide several examples below:

1. **portable version** - separate portable version- the program is packed in archive, the user extracts the archive and runs the executable, settings are saved in a configuration file in program folder;
2. **command-line argument** - if **crow.exe** is started with a specific parameter, for example `crow.exe -portable`, settings are saved in a configuration file in program folder;
3. **portable flag** - specific file in program folder (for example `portable` without extension or `portable.txt`) triggers portable mode, when **Crow Translate** starts it checks program folder for such file and if it is there settings are saved in a configuration file in program folder;
4. **configuration file** - when **Crow Translate** starts it checks program folder for a configuration file and if it is there the program uses it;
5. **special sub-folder** - when **Crow Translate** starts it checks program folder for a special sub-folder (for example `Data` or `Settings`) and if it is there it saves settings in a configuration file in this sub-folder;
6. **executable name** - if the executable is renamed from `crow.exe` to `crow-portable.exe` the program starts in portable mode, settings are saved in a configuration file in program folder;
7. **option in settings** - new check box- `Portable mode`, which, if selected, forces the program to save settings in a configuration file in program folder;
8. **portable installation** - when the program is installed beside normal installation it offers also portable installation (extraction), which just extracts the files to a folder specified by the user and settings are saved in a configuration file in program folder;

There may be other ways of triggering portable mode but I suppose the ones I suggested above are the most common. I prefer methods **1**, **3**, **4** and **5** but any of them would do once **Crow Translate** can save settings to configuration file and load settings from configuration file (and thus not using Windows Registry for saving/loading settings).

### Advantage
I think that portable programs (or programs running in portable mode) have several advantages:

- **mobility** - portable programs can be easily transferred from machine to machine with all their data and settings;
- **backup** - portable programs are easy to back up to external drives or to cloud services (in case one needs to re-install OS);
- **testing** - portable programs facilitate testing- one can test different versions (stable, alpha, beta) with different settings;

So I will be glad if you consider implementing portable mode for **Crow Translate** in future releases. I believe that if there is portable version of **Crow Translate** (or it can run in portable mode) this would add more flexibility to the already quite promising program.

**Comment by Shatur95** [comment]
Okay, I will add it :) Stay tuned for updates.

**Comment by Shatur95** [comment]
Done! I decided to combine the 4th and 7th example. You can create the file `settings.ini` and Crow will store the configuration in it. You can also open the app settings and enable "Portable mode" - the application will create and use this file itself.
Crow writes configuration only upon changes, so if you open the application for the first time, go to settings and enable portable mode, then no data will be written to the registry.

**Comment by smaragdus** [comment]
@Shatur95 

Excellent! I would appreciate any implementation of portable mode. The ones you chose are easy and handy for the user. Actually I wouldn't like very much only method 2- command-line argument, because the user needs to create a special shortcut or needs to use a program launcher which supports command-line arguments. Also the user may forget to start the program using the command-line parameter.

Do you plan to release Crow Translate as archive as well or you will provide it only as installer? In fact one can easily extract the installer (even 7-Zip was able to extract the installer right).

For me as a user it is always a pleasure to communicate with developers as helpful and as responsive like you.

Regards

**Comment by Shatur95** [comment]
Thanks, nice to hear :)

> Do you plan to release Crow Translate as archive as well or you will provide it only as installer? In fact one can easily extract the installer (even 7-Zip was able to extract the installer right).
Good idea, I think. I will publish the next release with archive too.

**Comment by smaragdus** [comment]
@Shatur95 

I downloaded version [2.2.3](https://github.com/crow-translate/crow-translate/releases/tag/2.2.3) but I neither found `Portable mode` check box in settings, nor the program used the file `settings.ini` I manually created in program folder.

Can you attach a screen shot which shows this `Portable mode` check box?

I created the configuration file (`settings.ini`) in the folder which contains the executable (`crow.exe`), this is the right place, right?

In short- I could not find the check box:

`Portable mode`

in settings.

The configuration file:

`settings.ini`

manually created in program folder is not used.

The settings are still saved in Windows Registry:

`HKEY_CURRENT_USER\Software\crow\Crow Translate`

I think that this issue needs to be re-opened.

**Comment by Shatur95** [comment]
Wow, it's my mistake :( I switched to automated releases and forgot to specify `PORTABLE_MODE` build flag in CI script. Stay tuned for updates.

**Comment by Shatur95** [comment]
Done, just wait until the binaries will be re-uploaded. You can see the progress [here](https://github.com/crow-translate/crow-translate/commit/d580e3828c3c6f6d199d4effa0b896390c04fd64/checks?check_suite_id=297026531).

**Comment by smaragdus** [comment]
@Shatur95 

Excellent! Portable mode really works as it is supposed to work, screen:

![Crow Translate 2 2 3 - 2019-11-05 - 002](https://user-images.githubusercontent.com/9025757/68243552-682b3300-001b-11ea-9ad7-191b68a46440.png)

When I checked `Portable mode` option the program created configuration file `settings.ini` in program folder.

However it seems that on first run the program still creates empty registry key:

`HKEY_CURRENT_USER\Software\Crow\Crow Translate`

So it seems that the better option is to manually create empty `settings.ini` in program folder.

Yet after I deleted the registry key it was not re-created on the next run.
